# CS203 - Software Tools and Techniques for AI
## Assignment - 3
Shah Harshil Hardik (23110132) <br>
Shah Tirth (23110295) <br>
Group No. 26 

### Import the necessary libraries

In [2]:
import pandas as pd
from sklearn.metrics import cohen_kappa_score
from statsmodels.stats.inter_rater import fleiss_kappa
from unidecode import unidecode

## Fleiss Kappa for CV annotations 

In [57]:
#Importing the memeber annotations data
mem1_cv = pd.read_json('JSON/project-1-at-2025-01-24-10-09-059bafe0.json')
mem2_cv = pd.read_json('JSON\project-3-at-2025-01-24-15-39-88aac066.json')
mem3_cv = pd.read_json('JSON\project-5-at-2025-01-24-19-29-36a76b78.json')

### Preprocessing data to remove unnecessary columns

In [56]:
#If truck spotted by annotator then truck column value is entered 1 else 0
def preprocess(df: pd.DataFrame):
    df.drop(columns=['annotator', 'annotation_id', 'created_at', 'updated_at', 'lead_time'], inplace=True)
    df['truck'] = df['choice'].apply(lambda x: 1 if x == 'Truck' else 0)

In [58]:
preprocess(mem1_cv)
preprocess(mem2_cv)
preprocess(mem3_cv)

In [59]:
print(mem1_cv)
print(mem2_cv)
print(mem3_cv)

                                  image  id    choice  truck
0   /data/upload/1/b99b4b3e-img_500.jpg   1  No Truck      0
1   /data/upload/1/ba99e33d-img_501.jpg   2  No Truck      0
2   /data/upload/1/ee510404-img_502.jpg   3  No Truck      0
3   /data/upload/1/c75f95aa-img_503.jpg   4     Truck      1
4   /data/upload/1/40f58a1c-img_504.jpg   5  No Truck      0
5   /data/upload/1/5eef8e28-img_505.jpg   6     Truck      1
6   /data/upload/1/f2cc5d00-img_506.jpg   7     Truck      1
7   /data/upload/1/bc21f97c-img_507.jpg   8  No Truck      0
8   /data/upload/1/fe362848-img_508.jpg   9  No Truck      0
9   /data/upload/1/400659c3-img_509.jpg  10     Truck      1
10  /data/upload/1/8d6eb3c3-img_510.jpg  11     Truck      1
11  /data/upload/1/563de9ba-img_511.jpg  12  No Truck      0
12  /data/upload/1/f8d62031-img_512.jpg  13     Truck      1
13  /data/upload/1/d7ab765b-img_513.jpg  14     Truck      1
14  /data/upload/1/97d6fffc-img_514.jpg  15     Truck      1
15  /data/upload/1/6a9f0

### Function to get data for calculating Fleiss Kappa in required format

In [60]:
#This function brings the data in the format required by the fleiss_kappa function
def get_data(df_list: list):
    data = []
    for i in range(len(df_list[0])):
        truck = no_truck = 0
        for j in range(len(df_list)):
            if (df_list[j]['truck'][i] == 1):
                truck += 1
            else:
                no_truck += 1
        data.append([truck, no_truck])
    return data

In [61]:
#Calculating the fleiss kappa based on the data acquired
data = get_data([mem1_cv, mem2_cv, mem3_cv])
fleiss_kappa_score = fleiss_kappa(data)
print(f"Fleiss Kappa = {fleiss_kappa_score}")

Fleiss Kappa = 0.8628571428571427


### Interpretation
A fleiss kappa value = 0.863 indicates a high level of agreement among the annotators. Moreover, it implies that there were only few confusing images which had caused diagreement between the annotators. 

## Cohen's Kappa for NLP

### Import the data

In [65]:
mem1_nlp = pd.read_json('JSON\project-4-at-2025-01-24-10-09-da3b5867.json')
mem2_nlp = pd.read_json('JSON\project-9-at-2025-01-24-22-40-c923da3e.json')

In [62]:
def preprocess(df: pd.DataFrame):
    df.drop(columns=['annotator', 'annotation_id', 'created_at', 'updated_at', 'lead_time'], inplace=True)

In [66]:
preprocess(mem1_nlp)
preprocess(mem2_nlp)

In [22]:
mem1_annotations = []
mem2_annotations = []
for i in range(len(mem1_nlp)):
    assert len(mem1_nlp['label'][i]) ==  len(mem2_nlp['label'][i])
    for j in range(len(mem1_nlp['label'][i])):
        mem1_annotations.append((unidecode(mem1_nlp['label'][i][j]['text']), mem1_nlp['label'][i][j]['labels']))
        mem2_annotations.append((unidecode(mem2_nlp['label'][i][j]['text']), mem2_nlp['label'][i][j]['labels']))


In [23]:
len(mem1_annotations), len(mem2_annotations)

(415, 415)

In [24]:
print(mem1_annotations)
print(mem2_annotations)

[('videsh', ['PROPN']), ('mNtraaly', ['PROPN']), (',', ['X']), ('Commerce', ['PROPN']), ('&', ['X']), ('Trade', ['PROPN']), ('aur', ['CONJ']), ('ASSOCHAM', ['PROPN']), ('jaise', ['ADP']), ('iNddsttrii', ['NOUN']), (' ke', ['ADP']), ('sNgtthnoN', ['NOUN']), ('ke', ['ADP']), ('biic', ['ADP']), ('behtr', ['ADJ']), ('taalmel', ['NOUN']), ('aaj', ['ADV']), ('smy', ['NOUN']), ('kii', ['ADP']), ('maaNg', ['NOUN']), ('hai', ['VERB']), (' / ', ['X']), ('purtgaal', ['PROPN']), ('Shri', ['PROPN']), ('Mario', ['PROPN']), (' Soares', ['PROPN']), (' ruup', ['NOUN']), ('mehmaan', ['NOUN']), ('gnntNtr', ['NOUN']), ('divs', ['NOUN']), ('netaa', ['NOUN']), ('pdhaare', ['VERB']), ('the', ['VERB']), (' / ', ['X']), ('ke', ['ADP']), ('meN ', ['ADP']), ('ke', ['ADP']), ('ke', ['ADP']), (' ke', ['ADP']), ('meN ', ['ADP']), ('mhaan', ['ADJ']), ('mukhy', ['ADJ']), ('1992', ['NUM']), ('bhaart', ['PROPN']), ('hmne', ['PROPN']), ('kaatthmaaNdduu', ['PROPN']), ('ko', ['ADP']), ('bhaart', ['PROPN']), ('se', ['ADP']

In [28]:
#Calculating the cohen's kappa based on the acquired data
l1 = [x[1] for x in mem1_annotations]
l2 = [x[1] for x in mem2_annotations]
cohen_kappa_score(l1, l2)

0.9324301404991757